### PR TITLE
Fix bug: updating potentially deleted CUDT

### PR DESCRIPTION
**Pull request by ethouris**
The CUDT entities that are failed are being updated in the `updateConnStatus`, but this is performed with no lock and there's a risk that the 1s time has passed between the moment when the GC thread decided to put the socket into the trashcan and when this thread is executed. Actually this 1s timer is the only certainty that the CUDT still exist provided it wasn't removed earlier. It has happened sometimes in the unit test that it crashed by accessing an already deleted object. This fix causes that the entity is first checked if its associated socket still exists and if not, it simply ignores this entity, as the pointer to it in this case is already dangling.

There are two problems with this fix:
1. This doesn't actually fix the problem. The problem is that the CUDT entity should be FIRST UNMAPPED from the container in the queue, and THEN DELETED, and likely all mapped CUDTs should be somehow flagged to not get deleted until this function finishes. This is then only a workaround that should ensure the time-related impossibility of deletion since the call to `locateSocket`.
2. This fix is also causing a mysterious hangup on Windows, which looks like a deadlock, however a frozen thread that attempts to lock the GlobControlLock doesn't seem to be frozen on that mutex, as it's not showing itself busy.

This PR presents the fix as described, but this should be further investigated.

### CODE REVIEW

**Review by maxsharabayko — CHANGES_REQUESTED** [review]
Given those two considerations, what would be the effort and the scope of changes for a proper fix?
I don't think we can or should merge something that causes a mysterious hangup.

> 1. This doesn't actually fix the problem. ...
> 2. This fix is also causing a mysterious hangup on Windows, ...

**Comment by maxsharabayko** [comment]
@ethouris  Closing?

**Comment by ethouris** [comment]
I'd still consider merging it as a sanity check. Even though the problem is otherwise fixed.

File: srtcore/queue.cpp [hunk 1]
```diff
@@ -1097,6 +1097,14 @@ void CRendezvousQueue::updateConnStatus(EReadStatus rst, EConnectStatus cst, con
     for (vector<LinkStatusInfo>::iterator i = ufailed.begin(); i != ufailed.end(); ++i)
     {
         HLOGC(cnlog.Debug, log << "updateConnStatus: COMPLETING dep objects update on failed @" << i->id);
+        // This is still done without locks, but a successful call to `locateSocket` should
+        // ensure that this entity will not be deleted in the next 1 second.
+        CUDTSocket* s = CUDT::uglobal()->locateSocket(i->id);
+        if (!s)
```
Comment (maxsharabayko) [inline srtcore/queue.cpp:1103]:
```suggestion
        if (CUDT::uglobal()->locateSocket(i->id) == NULL)
```

**Review by maxsharabayko — COMMENTED** [review]
> I'd still consider merging it as a sanity check. Even though the problem is otherwise fixed.

What about this statement then? Is it still true?

> This fix is also causing a mysterious hangup on Windows, which looks like a deadlock, however a frozen thread that attempts to lock the GlobControlLock doesn't seem to be frozen on that mutex, as it's not showing itself busy.

File: srtcore/queue.cpp (hunk 1 above)
Comment (ethouris) [reply srtcore/queue.cpp:1103]: I'll retest it on Windows first. When this still proves to be a problem, we can't merge it, even as sanity check.

**Comment by ethouris** [comment]
Too little to improve anything and brings a potential hangup risk.

**Pull request closed without merging**